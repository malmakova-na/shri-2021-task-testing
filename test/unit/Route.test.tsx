import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import events from "@testing-library/user-event";
import { Router } from "react-router";
import { Provider } from "react-redux";
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';


import { Application } from "../../src/client/Application";
import { ExampleApi, CartApi } from "../../src/client/api";
import { initStore } from "../../src/client/store";

describe('Переход по адресу: ', function() {
    it('/delivery открывается страница "Delivery"', () => {
        const history = createMemoryHistory({
        initialEntries: ['/delivery'],
        initialIndex: 0
        });

        const basename = "/hw/store";

        const api = new ExampleApi(basename);
        const cart = new CartApi();
        const store = initStore(api, cart);

        const application = (
        <Router history={history}>
            <Provider store={store}>
                <Application />
            </Provider>
        </Router>
        );
        const {getByRole} = render(application);
        events.click(screen.getByRole('link', {
            name: /catalog/i
        }));
        expect(
            screen.getByRole('heading', { name: /catalog/i}).textContent
        ).toEqual('Catalog');
        const contactSnapshot = renderer.create(application).toJSON();
      
        expect(contactSnapshot).toMatchSnapshot();
    });

    it('/ открывается главная страница приложения', () => {
        const history = createMemoryHistory({
        initialEntries: ['/'],
        initialIndex: 0
        });
        
        const basename = "/hw/store";
        
        const api = new ExampleApi(basename);
        const cart = new CartApi();
        const store = initStore(api, cart);
        
        const application = (
        <Router history={history}>
            <Provider store={store}>
                <Application />
            </Provider>
        </Router>
        );
        const {getByRole} = render(application);
        expect(
            screen.getByText(/welcome to example store!/i)
        ).toBeInTheDocument();
        const appSnapshot = renderer.create(application).toJSON();
      
        expect(appSnapshot).toMatchSnapshot();
        //screen.logTestingPlaygroundURL();
    });

    it('/cart открывается страница "Cart"', () => {
        const history = createMemoryHistory({
        initialEntries: ['/cart'],
        initialIndex: 0
        });
        
        const basename = "/hw/store";
        
        const api = new ExampleApi(basename);
        const cart = new CartApi();
        const store = initStore(api, cart);
        
        const application = (
        <Router history={history}>
            <Provider store={store}>
                <Application />
            </Provider>
        </Router>
        );
        const {getByRole, container} = render(application);
        const cartSnapshot = renderer.create(application).toJSON();
        
        events.click(screen.getByRole('link', {
            name: /cart/i
        }));
        expect(cartSnapshot).toMatchSnapshot();
        expect(
            screen.getByRole('heading', { name: /cart/i}).textContent
        ).toEqual('Shopping cart');
        
    });
    it('/contacts открывается страница "Contacts"', () => {
        const history = createMemoryHistory({
        initialEntries: ['/contacts'],
        initialIndex: 0
        });
        
        const basename = "/hw/store";
        
        const api = new ExampleApi(basename);
        const cart = new CartApi();
        const store = initStore(api, cart);
        
        const application = (
        <Router history={history}>
            <Provider store={store}>
                <Application />
            </Provider>
        </Router>
        );
        const {getByRole} = render(application);
        events.click(screen.getByRole('link', {
            name: /contacts/i
        }));
        expect(
            screen.getByRole('heading', { name: /contacts/i}).textContent
        ).toEqual('Contacts');
        const contactSnapshot = renderer.create(application).toJSON();
      
        expect(contactSnapshot).toMatchSnapshot();
    });

    it('/catalog открывается страница "Catalog"', () => {
        const history = createMemoryHistory({
        initialEntries: ['/catalog'],
        initialIndex: 0
        });
        
        const basename = "/hw/store";
        
        const api = new ExampleApi(basename);
        const cart = new CartApi();
        const store = initStore(api, cart);
        
        const application = (
        <Router history={history}>
            <Provider store={store}>
                <Application />
            </Provider>
        </Router>
        );
        const {getByRole} = render(application);
        events.click(screen.getByRole('link', {
            name: /catalog/i
        }));
        expect(
            screen.getByRole('heading', { name: /catalog/i}).textContent
        ).toEqual('Catalog');
    });
});
