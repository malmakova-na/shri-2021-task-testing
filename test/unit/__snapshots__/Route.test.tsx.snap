// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Переход по адресу:  / открывается главная страница приложения 1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/cart"
            onClick={[Function]}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Home"
    >
      <div
        className="row"
      >
        <div
          className="col bg-secondary text-white py-4 bg-opacity-75"
        >
          <p
            className="display-3"
          >
            Welcome to Example store!
          </p>
          <p
            className="lead"
          >
            Culpa perspiciatis corporis facilis fugit similique
          </p>
          <p
            className="lead"
          >
            Cum aut ut eveniet rem cupiditate natus veritatis quia
          </p>
        </div>
      </div>
      <div
        className="row mb-4"
      >
        <div
          className="col-12 col-md-4 bg-light py-3"
        >
          <h1>
            Quickly
          </h1>
          <p
            className="lead"
          >
            Odio aut assumenda ipsam amet reprehenderit. Perspiciatis qui molestiae qui tempora quisquam
          </p>
        </div>
        <div
          className="col-12 col-md-4 bg-light py-3"
        >
          <h1>
            Qualitatively
          </h1>
          <p
            className="lead"
          >
            Ut nisi distinctio est non voluptatem. Odio aut assumenda ipsam amet reprehenderit
          </p>
        </div>
        <div
          className="col-12 col-md-4 bg-light py-3"
        >
          <h1>
            Inexpensive
          </h1>
          <p
            className="lead"
          >
            Perspiciatis qui molestiae qui tempora quisquam. Ut nisi distinctio est non voluptatem
          </p>
        </div>
      </div>
      <div
        className="row mb-4"
      >
        <div
          className="col-12py-3"
        >
          <p>
            Sed voluptatum quis voluptates laudantium incidunt laudantium. Illo non quos eos vel ipsa. Explicabo itaque est optio neque rerum provident enim qui sed. Corrupti commodi voluptatem vero soluta hic.
          </p>
          <p>
            Modi corporis consectetur aliquid sit cum tenetur enim. Sed voluptatum quis voluptates laudantium incidunt laudantium. Illo non quos eos vel ipsa. Explicabo itaque est optio neque rerum provident enim qui sed. Corrupti commodi voluptatem vero soluta hic.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Переход по адресу:  /cart открывается страница "Cart" 1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current="page"
            className="nav-link active"
            href="/cart"
            onClick={[Function]}
            style={Object {}}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Cart"
    >
      <div
        className="row mb-4"
      >
        <div
          className="col"
        >
          <h1>
            Shopping cart
          </h1>
          Cart is empty. Please select products in the 
          <a
            href="/catalog"
            onClick={[Function]}
          >
            catalog
          </a>
          .
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Переход по адресу:  /catalog открывается страница "Catalog" 1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current="page"
            className="nav-link active"
            href="/catalog"
            onClick={[Function]}
            style={Object {}}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/cart"
            onClick={[Function]}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Catalog"
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <h1>
            Catalog
          </h1>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12 col-sm-6 col-md-4 col-lg-3"
          data-testid={0}
        >
          <div
            className="ProductItem card w-100 mb-4"
            data-testid={0}
          >
            <img
              className="Image card-img-top"
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
            />
            <div
              className="card-body"
            >
              <h5
                className="ProductItem-Name card-title"
              >
                test-product-name
              </h5>
              <p
                className="ProductItem-Price card-text"
              >
                $
                1
              </p>
              <a
                className="ProductItem-DetailsLink card-link"
                href="/catalog/0"
                onClick={[Function]}
              >
                Details
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-12 col-sm-6 col-md-4 col-lg-3"
          data-testid={1}
        >
          <div
            className="ProductItem card w-100 mb-4"
            data-testid={1}
          >
            <img
              className="Image card-img-top"
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
            />
            <div
              className="card-body"
            >
              <h5
                className="ProductItem-Name card-title"
              >
                test-product-name
              </h5>
              <p
                className="ProductItem-Price card-text"
              >
                $
                2
              </p>
              <a
                className="ProductItem-DetailsLink card-link"
                href="/catalog/1"
                onClick={[Function]}
              >
                Details
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-12 col-sm-6 col-md-4 col-lg-3"
          data-testid={2}
        >
          <div
            className="ProductItem card w-100 mb-4"
            data-testid={2}
          >
            <img
              className="Image card-img-top"
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
            />
            <div
              className="card-body"
            >
              <h5
                className="ProductItem-Name card-title"
              >
                test-product-name
              </h5>
              <p
                className="ProductItem-Price card-text"
              >
                $
                3
              </p>
              <a
                className="ProductItem-DetailsLink card-link"
                href="/catalog/2"
                onClick={[Function]}
              >
                Details
              </a>
            </div>
          </div>
        </div>
        <div
          className="col-12 col-sm-6 col-md-4 col-lg-3"
          data-testid={3}
        >
          <div
            className="ProductItem card w-100 mb-4"
            data-testid={3}
          >
            <img
              className="Image card-img-top"
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
            />
            <div
              className="card-body"
            >
              <h5
                className="ProductItem-Name card-title"
              >
                test-product-name
              </h5>
              <p
                className="ProductItem-Price card-text"
              >
                $
                4
              </p>
              <a
                className="ProductItem-DetailsLink card-link"
                href="/catalog/3"
                onClick={[Function]}
              >
                Details
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Переход по адресу:  /contacts открывается страница "Contacts" 1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current="page"
            className="nav-link active"
            href="/contacts"
            onClick={[Function]}
            style={Object {}}
          >
            Contacts
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/cart"
            onClick={[Function]}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Contacts"
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <h1>
            Contacts
          </h1>
          <p>
            Ut non consequatur aperiam ex dolores. Voluptatum harum consequatur est totam. Aut voluptatum aliquid aut optio et ea. Quaerat et eligendi minus quasi. Culpa voluptatem voluptatem dolores molestiae aut quos iure. Repellat aperiam ut aliquam iure. Veritatis magnam quisquam et dolorum recusandae aut.
          </p>
          <p>
            Molestias inventore illum architecto placeat molestias ipsam facilis ab quo. Rem dolore cum qui est reprehenderit assumenda voluptatem nisi ipsa. Unde libero quidem. Excepturi maiores vel quia. Neque facilis nobis minus veniam id. Eum cum eveniet accusantium molestias voluptas aut totam laborum aut. Ea molestiae ullam et. Quis ea ipsa culpa eligendi ab sit ea error suscipit. Quia ea ut minus distinctio quam eveniet nihil. Aut voluptate numquam ipsa dolorem et quas nemo.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Переход по адресу:  /delivery открывается страница "Delivery" 1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            aria-current="page"
            className="nav-link active"
            href="/delivery"
            onClick={[Function]}
            style={Object {}}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/cart"
            onClick={[Function]}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Delivery"
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <h1>
            Delivery
          </h1>
          <p>
            Deserunt occaecati tempora. Qui occaecati est aliquam. Enim qui nulla ipsam. Incidunt impedit enim consequuntur amet at consequuntur vero. Dolor et ad facere asperiores iste est praesentium quaerat iure. Quibusdam mollitia autem quos voluptas quia est doloremque corporis et. Sed fuga quasi esse perspiciatis fugit maxime. Qui quidem amet.
          </p>
          <img
            className="Image w-25 mb-4"
            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
          />
          <p>
            Dolores magnam consequatur iste aliquam qui sint non ab. Culpa saepe omnis. Recusandae vel aperiam voluptates harum. Perspiciatis qui molestiae qui tempora quisquam. Mollitia voluptatum minus laboriosam. Dolor maiores possimus repudiandae praesentium hic eos. Veritatis et repellat.
          </p>
          <p>
            Pariatur nisi nobis hic ut facilis sunt rerum id error. Soluta nihil quisquam quia rerum illo. Ipsam et suscipit est iure incidunt quasi et eum. Culpa libero dignissimos recusandae. In magni sapiente non voluptas molestias. Deserunt quos quo placeat sunt. Ea necessitatibus dolores eaque ex aperiam sunt eius. Saepe aperiam aut. Quaerat natus consequatur aut est id saepe et aut facilis.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
